From 66d300a59229a2eea73930fbdd5a8d076c44efe1 Mon Sep 17 00:00:00 2001
From: bhavya <bhavya25hirani@gmail.com>
Date: Wed, 5 Mar 2025 09:35:05 -0500
Subject: [PATCH 1/3] working build with support to inline classes specified in
 a regular expressions

for JCL classes that are specified like so, all its ancestors in the class hierarchy, leaving Object, must be mentioned
---
 runtime/bcutil/ClassFileOracle.cpp     | 58 ++++++++++++++++++++++++-
 runtime/bcutil/ClassFileOracle.hpp     | 24 +++++++++++
 runtime/bcutil/ROMClassBuilder.cpp     | 14 ++++++
 runtime/gc_modron_startup/mgcalloc.cpp | 49 ++++++++++++++++++++-
 runtime/oti/j9.h                       |  1 +
 runtime/oti/j9modifiers_api.h          |  3 ++
 runtime/util/resolvehelp.c             |  2 +
 runtime/vm/ObjectMonitor.cpp           |  3 ++
 runtime/vm/createramclass.cpp          | 59 ++++++++++++++++++++++++--
 9 files changed, 207 insertions(+), 6 deletions(-)

diff --git a/runtime/bcutil/ClassFileOracle.cpp b/runtime/bcutil/ClassFileOracle.cpp
index 359dea5c20..93e1ce3853 100644
--- a/runtime/bcutil/ClassFileOracle.cpp
+++ b/runtime/bcutil/ClassFileOracle.cpp
@@ -35,6 +35,8 @@
 #include "ut_j9bcu.h"
 #include "util_api.h"
 #include "j9protos.h"
+//inliningjclclasses
+#include<iostream>
 
 /* The array entries must be in same order as the enums in ClassFileOracle.hpp */
 ClassFileOracle::KnownAnnotation ClassFileOracle::_knownAnnotations[] = {
@@ -368,6 +370,27 @@ ClassFileOracle::walkFields()
 
 		markConstantUTF8AsReferenced(field->nameIndex);
 		markConstantUTF8AsReferenced(field->descriptorIndex);
+		
+		//inliningjclclasses
+
+		if((isFlattenablePrimitiveClassBH((char *)(this->getUTF8Data(field->descriptorIndex))) && !isLibraryClassBH((char *)(this->getUTF8Data(this->getClassNameIndex()))))){
+			std::cerr<<"FLATTENABLE IN A NON LIBRARY CLASS, CONTAINING CLASS IS: "<<((char *)(this->getUTF8Data(this->getClassNameIndex())))<<"\n";
+			_fieldsInfo[fieldIndex].isNullRestricted = true;
+		
+			std::cerr<<"FIELD VARIABLE NAME: "<<((char *)(this->getUTF8Data(field->nameIndex)))<<"\n";
+
+		}else{
+			if(isLibraryClassBH((char *)(this->getUTF8Data(this->getClassNameIndex()))))
+			{
+				//std::cout<<"LIBRARY CLASS HENCE AVOIDING NULLRESTRICTED SET: "<<((char *)(this->getUTF8Data(this->getClassNameIndex())))<<"\n";
+			}
+
+		}
+
+		/*if((isFlattenablePrimitiveClassBH((char *)(this->getUTF8Data(field->descriptorIndex))))){
+			std::cerr<<"FLATTENABLE FIELD BEING SET AS NULL RESTRICTED, CONTAINING CLASS IS: "<<((char *)(this->getUTF8Data(this->getClassNameIndex())))<<"\n";
+			_fieldsInfo[fieldIndex].isNullRestricted = true;
+		}*/
 
 		if (isStatic) {
 			if (NULL != field->constantValueAttribute) {
@@ -410,6 +433,8 @@ ClassFileOracle::walkFields()
 				_fieldsInfo[fieldIndex].isSynthetic = true;
 				break;
 			case CFR_ATTRIBUTE_RuntimeVisibleAnnotations: {
+
+				//std::cout<< "FOUND RUNTIMEVISIBLEANNOTATIONS IN A FIELD!! It's Descriptor Is:" << (char *)(this->getUTF8Data(field->descriptorIndex)) << "!!\n";
 				J9CfrAttributeRuntimeVisibleAnnotations *attribAnnotations = (J9CfrAttributeRuntimeVisibleAnnotations *)attrib;
 				UDATA knownAnnotations = 0;
 				if ((NULL != _context->javaVM()) && (J9_ARE_ALL_BITS_SET(_context->javaVM()->extendedRuntimeFlags, J9_EXTENDED_RUNTIME_ALLOW_CONTENDED_FIELDS)) &&
@@ -426,7 +451,9 @@ ClassFileOracle::walkFields()
 						_fieldsInfo[fieldIndex].isFieldContended = true;
 					}
 #if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+					//inliningjclclasses
 					if (containsKnownAnnotation(foundAnnotations, NULLRESTRICTED_ANNOTATION)) {
+					//std::cout<< "FOUND NULL RESTRICTED FIELD!! It's Descriptor Is:" << (char *)(this->getUTF8Data(field->descriptorIndex)) << "!!\n";
 						if (!IS_CLASS_SIGNATURE(fieldChar)) {
 							if ('[' == fieldChar) {
 								throwGenericErrorWithCustomMsg(J9NLS_CFR_NO_NULLRESTRICTED_IN_ARRAYFIELD__ID, fieldIndex);
@@ -483,6 +510,33 @@ ClassFileOracle::walkAttributes()
 {
 	ROMClassVerbosePhase v(_context, ClassFileAttributesAnalysis);
 
+	//inliningjclclasses: if we want this class to be inlined, implicit creation flags must be turned on
+	if(isFlattenablePrimitiveClassBH((char *)(this->getUTF8Data(this->getClassNameIndex()))))
+	{
+		std::cerr<<"SETTING IMPLICITCREATEHASDEFAULTVALUE IN CLASS: "<< ((char *)(this->getUTF8Data(this->getClassNameIndex())))<<"\n";
+		_hasImplicitCreationAttribute = true;
+		_implicitCreationFlags |= J9AccImplicitCreateHasDefaultValue;
+	}
+	// inliningjclclasses: setting identity to false if we want this class to get inlined. 
+	// "value" is determined by a lack of identity flag in the current openj9 version
+	if(isFlattenablePrimitiveClassBH((char *)(this->getUTF8Data(this->getClassNameIndex()))))
+	{
+		std::cerr<<"TURNING OFF IDENTITY IN CLASS: "<< ((char *)(this->getUTF8Data(this->getClassNameIndex())))<<"\n";
+		//(_classFile->accessFlags) &= ~CFR_ACC_IDENTITY;
+		(_classFile->accessFlags) &= ~J9AccClassHasIdentity;
+		(_classFile->accessFlags) &= ~J9AccInterface;
+	}
+
+	// inliningjclclasses: setting the right classfile version:
+	if(isFlattenablePrimitiveClassBH((char *)(this->getUTF8Data(this->getClassNameIndex()))))
+	{
+		std::cerr<<"SETTING RIGHT VERSION IN CLASSFILE: "<< ((char *)(this->getUTF8Data(this->getClassNameIndex())))<<"\n";
+		std::cerr<<"MAJOR VERSION: "<<_classFile->majorVersion<<", MINOR VERSION: "<< _classFile->minorVersion << "\n";
+		//_classFile->majorVersion = 68;
+		_classFile->minorVersion = 65535;
+	}
+	
+
 	for (U_16 attributeIndex = 0; attributeIndex < _classFile->attributesCount; attributeIndex++) {
 		J9CfrAttribute *attrib = _classFile->attributes[attributeIndex];
 		switch (attrib->tag) {
@@ -578,9 +632,11 @@ ClassFileOracle::walkAttributes()
 				}
 				if (containsKnownAnnotation(foundAnnotations, VALUEBASED_ANNOTATION)) {
 					_isClassValueBased = true;
+					//std::cout<< "FOUND VALUE BASED CLASS!! It's " << (char *)(this->getUTF8Data(this->getClassNameIndex())) << "!!\n";
 				}
 #if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
-				if (containsKnownAnnotation(foundAnnotations, IMPLICITLYCONSTRUCTIBLE_ANNOTATION)) {
+				// inliningjclclasses: added an OR with VALUEBASED_ANNOTATION
+				if (containsKnownAnnotation(foundAnnotations, IMPLICITLYCONSTRUCTIBLE_ANNOTATION) ) {// || containsKnownAnnotation(foundAnnotations, VALUEBASED_ANNOTATION)) {
 					_hasImplicitCreationAttribute = true;
 					_implicitCreationFlags |= J9AccImplicitCreateHasDefaultValue;
 				}
diff --git a/runtime/bcutil/ClassFileOracle.hpp b/runtime/bcutil/ClassFileOracle.hpp
index 46854d04c0..9fecdebb15 100644
--- a/runtime/bcutil/ClassFileOracle.hpp
+++ b/runtime/bcutil/ClassFileOracle.hpp
@@ -34,6 +34,9 @@
 #include "ut_j9bcu.h"
 #include "bcnames.h"
 
+//inliningjclclasses
+#include<regex>
+
 #include "BuildResult.hpp"
 #if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
 #include "VMHelpers.hpp"
@@ -927,6 +930,27 @@ class RecordComponentIterator
 	/*
 	 * Query methods.
 	 */
+	//inliningjclclasses
+	bool isLibraryClassBH(char *descriptor) const {
+		static const std::regex prefix_regex(R"(^(?:java/|sun/|javax/|com/sun/|org/omg/|org/xml/|org/w3c/dom/|openj9/internal/|build/))");
+		return std::regex_search(reinterpret_cast<const char*>(descriptor), prefix_regex);
+	}
+
+	bool isFlattenablePrimitiveClassBH(char *descriptor) const {
+		static const std::regex prefix_regex(R"(^(?:L)?(?:java/lang/Integer|InlineField|java/lang/Number)(;)?$)");
+		return std::regex_search(reinterpret_cast<const char*>(descriptor), prefix_regex);
+	}
+
+	bool isCustomFlattenablePrimitiveClassBH(char *descriptor) const {
+		static const std::regex prefix_regex(R"(^(?:L)?(?:InlineField)(;)?$)");
+		return std::regex_search(reinterpret_cast<const char*>(descriptor), prefix_regex);
+	}
+
+	/*bool isFlattenablePrimitiveClassBH(char *descriptor) const {
+		static const std::regex prefix_regex(R"(^(?:L)?(?:java/lang/Integer)(;)?$)");
+		return false;
+		//return std::regex_search(reinterpret_cast<const char*>(descriptor), prefix_regex);
+	}*/
 
 	U_32 getClassFileSize() const { return _classFile->classFileSize; }
 	U_16 getAccessFlags() const { return _classFile->accessFlags; }
diff --git a/runtime/bcutil/ROMClassBuilder.cpp b/runtime/bcutil/ROMClassBuilder.cpp
index 16230105b0..3b100046b8 100644
--- a/runtime/bcutil/ROMClassBuilder.cpp
+++ b/runtime/bcutil/ROMClassBuilder.cpp
@@ -51,6 +51,8 @@
 #include "WritingCursor.hpp"
 #include "j9protos.h"
 #include "ut_j9bcu.h"
+//inliningjclclasses
+#include<iostream>
 
 static const UDATA INITIAL_CLASS_FILE_BUFFER_SIZE = 4096;
 static const UDATA INITIAL_BUFFER_MANAGER_SIZE = 32768 * 10;
@@ -584,6 +586,18 @@ ROMClassBuilder::prepareAndLaydown( BufferManager *bufferManager, ClassFileParse
 	romClassWriter.setSRPOffsetTable(&srpOffsetTable);
 
 	U_32 modifiers = classFileOracle.getAccessFlags();
+
+	//inliningjclclasses
+	if(classFileOracle.isFlattenablePrimitiveClassBH((char *)(classFileOracle.getUTF8Data(classFileOracle.getClassNameIndex()))))
+	{
+		//std::cout<<"TURNING OFF IDENTITY IN CLASS: "<< ((char *)(this->getUTF8Data(this->getClassNameIndex())))<<"\n";
+		//(_classFile->accessFlags) &= ~CFR_ACC_IDENTITY;
+		if(!J9_ARE_ALL_BITS_SET(modifiers, CFR_ACC_IDENTITY))
+		{
+			std::cout<<"THE FOLLOWING CLASS HAS NO IDENTITY IN ROMCLASSBUILDER: "<< ((char *)(classFileOracle.getUTF8Data(classFileOracle.getClassNameIndex())))<<"\n";
+		}
+	}
+
 	U_32 extraModifiers = computeExtraModifiers(&classFileOracle, context);
 	U_32 optionalFlags = computeOptionalFlags(&classFileOracle, context);
 
diff --git a/runtime/gc_modron_startup/mgcalloc.cpp b/runtime/gc_modron_startup/mgcalloc.cpp
index 334c1639e6..26cc2b2560 100644
--- a/runtime/gc_modron_startup/mgcalloc.cpp
+++ b/runtime/gc_modron_startup/mgcalloc.cpp
@@ -54,6 +54,8 @@
 #include "Scheduler.hpp"
 #endif /* J9VM_GC_REALTIME */
 #include "VMAccess.hpp"
+//inliningjclclasses
+#include<iostream>
 
 extern "C" {
 
@@ -578,6 +580,8 @@ J9AllocateIndexableObject(J9VMThread *vmThread, J9Class *clazz, uint32_t numberO
 		Assert_MM_true(GC_ObjectModel::SCAN_PRIMITIVE_ARRAY_OBJECT == extensions->objectModel.getScanType(clazz));
 	}
 
+	//inliningjclclasses comment
+	//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : ONE"<<"\n";
 #if defined(J9VM_GC_THREAD_LOCAL_HEAP)
 	if (!env->isInlineTLHAllocateEnabled()) {
 		/* For duration of call restore TLH allocate fields;
@@ -588,6 +592,8 @@ J9AllocateIndexableObject(J9VMThread *vmThread, J9Class *clazz, uint32_t numberO
 	}	
 #endif /* J9VM_GC_THREAD_LOCAL_HEAP */
 
+	//inliningjclclasses comment
+	//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : TWO"<<"\n";
 	J9Object *objectPtr = NULL;
 	uintptr_t sizeInBytesRequired = 0;
 	MM_IndexableObjectAllocationModel indexableOAM(env, clazz, numberOfIndexedFields, allocateFlags);
@@ -604,6 +610,8 @@ J9AllocateIndexableObject(J9VMThread *vmThread, J9Class *clazz, uint32_t numberO
 		}
 	}
 	
+	//inliningjclclasses comment
+	//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : THREE"<<"\n";
 	if (env->_failAllocOnExcessiveGC && (NULL != objectPtr)) {
 		/* If we have garbage collected too much, return NULL as if we had failed to allocate the object (effectively triggering an OOM).
 		 * TODO: The ordering of this call wrt/ allocation really needs to change - this is just a temporary solution until
@@ -622,6 +630,8 @@ J9AllocateIndexableObject(J9VMThread *vmThread, J9Class *clazz, uint32_t numberO
 		Trc_MM_ArrayObjectAllocationFailedDueToExcessiveGC(vmThread);
 	}
 
+	//inliningjclclasses comment
+	//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : FOUR"<<"\n";
 	sizeInBytesRequired = indexableOAM.getAllocateDescription()->getBytesRequested();
 	if (NULL != objectPtr) {
 		/* The hook could release access and so the object address could change (the value is preserved).  Since this
@@ -646,6 +656,8 @@ J9AllocateIndexableObject(J9VMThread *vmThread, J9Class *clazz, uint32_t numberO
 					sizeInBytesRequired);
 			}
 		}
+		//inliningjclclasses comment
+		//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : FOUR ONE"<<"\n";
 	
 		/* If this was a non-TLH allocation, trigger the hook */
 		if( !indexableOAM.getAllocateDescription()->isCompletedFromTlh()) {
@@ -655,6 +667,8 @@ J9AllocateIndexableObject(J9VMThread *vmThread, J9Class *clazz, uint32_t numberO
 				objectPtr);
 		}
 		
+		//inliningjclclasses comment
+		//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : FOUR TWO"<<"\n";
 		uintptr_t lowThreshold = extensions->lowAllocationThreshold;
 		uintptr_t highThreshold = extensions->highAllocationThreshold;
 		if ( (sizeInBytesRequired >= lowThreshold) && (sizeInBytesRequired <= highThreshold) ) {
@@ -668,14 +682,23 @@ J9AllocateIndexableObject(J9VMThread *vmThread, J9Class *clazz, uint32_t numberO
 				highThreshold);
 		}
 		
+		//inliningjclclasses comment
+		//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : FOUR THREE"<<"\n";
 		objectPtr = traceAllocateObject(vmThread, objectPtr, clazz, sizeInBytesRequired, (uintptr_t)numberOfIndexedFields);
 		if (extensions->isStandardGC()) {
+
+			//inliningjclclasses comment
+			//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : FOUR FOUR"<<"\n";
 			if (OMR_GC_ALLOCATE_OBJECT_TENURED == (allocateFlags & OMR_GC_ALLOCATE_OBJECT_TENURED)) {
 				/* Object must be allocated in Tenure if it is requested */
 				Assert_MM_true(extensions->isOld(objectPtr));
 			}
+			//inliningjclclasses
+			//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : FOUR FIVE"<<"\n";
 #if defined(J9VM_GC_REALTIME)
 		} else if (extensions->isMetronomeGC()) {
+			//inliningjclclasses comment
+			//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : FOUR SIX"<<"\n";
 			if (env->saveObjects((omrobjectptr_t)objectPtr)) {
 				j9gc_startGCIfTimeExpired(vmThread->omrVMThread);
 				env->restoreObjects((omrobjectptr_t*)&objectPtr);
@@ -692,14 +715,36 @@ J9AllocateIndexableObject(J9VMThread *vmThread, J9Class *clazz, uint32_t numberO
 	}
 	/* TODO: Need to implement a more optimal path for cases where barriers are not required or where a batch barrier can be used. */ 
 	if ((NULL != objectPtr) && J9_ARE_ALL_BITS_SET(clazz->classFlags, J9ClassContainsUnflattenedFlattenables)) {
+		//inliningjclclasses comment
+		//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : FOUR SEVEN"<<"\n";
+
 		MM_ObjectAccessBarrierAPI objectAccessBarrier(vmThread);
 		J9Class * elementClass = ((J9ArrayClass *) clazz)->componentType; 
-		j9object_t defaultValue = elementClass->flattenedClassCache->defaultValue;
+
+		//inliningjclclasses comment
+		//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : FOUR EIGHT"<<"\n";
+
+		auto tempBH = elementClass->flattenedClassCache;
+
+		//j9object_t defaultValue = elementClass->flattenedClassCache->defaultValue;
+		std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : FOUR NINE"<<"\n";
+
+		//inliningjclclasses if condition for error check
+		if(tempBH == NULL)
+		{
+			//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : flattened class cache IS NULL!!"<<"\n";	
+		}	
+		
+		j9object_t defaultValue = tempBH->defaultValue;
+
+		//inliningjclclasses comment
+		//std::cerr<<"INSIDE J9ALLOCATEINDEXABLEOBJECT : FOUR TEN"<<"\n";
 		for (UDATA index = 0; index < numberOfIndexedFields; index++) {
 			objectAccessBarrier.inlineIndexableObjectStoreObject(vmThread, objectPtr, index, defaultValue);
 		}
-	}
 
+	}
+	
 #if defined(J9VM_GC_THREAD_LOCAL_HEAP)
 	if (extensions->needDisableInlineAllocation()) {
 		env->disableInlineTLHAllocate();
diff --git a/runtime/oti/j9.h b/runtime/oti/j9.h
index 682664d058..b6371dda3c 100644
--- a/runtime/oti/j9.h
+++ b/runtime/oti/j9.h
@@ -345,6 +345,7 @@ static const struct { \
 /**
  * Disable flattening of volatile field that is > 8 bytes for now, as the current implementation of copyObjectFields() will tear this field.
  */
+// inliningjclclasses: removed null restricted condition for J9_IS_FIELD_FLATTENED
 #define J9_IS_FIELD_FLATTENED(fieldClazz, romFieldShape) \
 		(J9ROMFIELD_IS_NULL_RESTRICTED(romFieldShape) && \
 		J9_IS_J9CLASS_FLATTENED(fieldClazz) && \
diff --git a/runtime/oti/j9modifiers_api.h b/runtime/oti/j9modifiers_api.h
index 086af962ed..c8794fe00e 100644
--- a/runtime/oti/j9modifiers_api.h
+++ b/runtime/oti/j9modifiers_api.h
@@ -82,7 +82,10 @@
 #define J9ROMCLASS_IS_CONTENDED(romClass)	_J9ROMCLASS_J9MODIFIER_IS_SET((romClass), J9AccClassIsContended)
 
 #ifdef J9VM_OPT_VALHALLA_VALUE_TYPES
+//inliningjclclasses: removing version check from J9ROMCLASS_IS_VALUE
 #define J9ROMCLASS_IS_VALUE(romClass)   (J9_IS_CLASSFILE_OR_ROMCLASS_VALUETYPE_VERSION(romClass) && !_J9ROMCLASS_SUNMODIFIER_IS_ANY_SET((romClass), J9AccClassHasIdentity | J9AccInterface))
+//#define J9ROMCLASS_IS_VALUE(romClass)   (!_J9ROMCLASS_SUNMODIFIER_IS_ANY_SET((romClass), J9AccClassHasIdentity | J9AccInterface))
+
 #else /* J9VM_OPT_VALHALLA_VALUE_TYPES */
 #define J9ROMCLASS_IS_VALUE(romClass)	FALSE
 #endif /* J9VM_OPT_VALHALLA_VALUE_TYPES */
diff --git a/runtime/util/resolvehelp.c b/runtime/util/resolvehelp.c
index 537820784a..90344a7a94 100644
--- a/runtime/util/resolvehelp.c
+++ b/runtime/util/resolvehelp.c
@@ -41,6 +41,8 @@ getMethodForSpecialSend(J9VMThread *vmStruct, J9Class *currentClass, J9Class *re
 	 *	C) Resolved method is not <init>
 	 *	D) Skip checking vTables if resolved or current class is an interface
 	 */
+
+	//inliningjclclasses : attempting to comment out the J9AccSuper condition as per point (A) above
 	if (
 #if !defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
 		J9_ARE_ALL_BITS_SET(currentClass->romClass->modifiers, J9AccSuper) ||
diff --git a/runtime/vm/ObjectMonitor.cpp b/runtime/vm/ObjectMonitor.cpp
index 2e903c1285..d298e02e45 100644
--- a/runtime/vm/ObjectMonitor.cpp
+++ b/runtime/vm/ObjectMonitor.cpp
@@ -328,12 +328,14 @@ objectMonitorEnterNonBlocking(J9VMThread *currentThread, j9object_t object)
 #endif /* defined(J9VM_OPT_CRIU_SUPPORT) */
 
 #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+	// inliningjclclasses: establishing equivalence between valuetype and valuebased
 	if (J9_IS_J9CLASS_VALUETYPE(objClass)) {
 		result = J9_OBJECT_MONITOR_VALUE_TYPE_IMSE;
 		goto done;
 	}
 #endif /* J9VM_OPT_VALHALLA_VALUE_TYPES */
 #if JAVA_SPEC_VERSION >= 16
+	
 	if (J9_IS_J9CLASS_VALUEBASED(objClass)) {
 		U_32 runtimeFlags2 = vm->extendedRuntimeFlags2;
 		if (J9_ARE_ALL_BITS_SET(runtimeFlags2, J9_EXTENDED_RUNTIME2_VALUE_BASED_EXCEPTION)) {
@@ -345,6 +347,7 @@ objectMonitorEnterNonBlocking(J9VMThread *currentThread, j9object_t object)
 			j9nls_printf(PORTLIB, J9NLS_WARNING, J9NLS_VM_ERROR_BYTECODE_OBJECTREF_CANNOT_BE_VALUE_BASED, J9UTF8_LENGTH(className), J9UTF8_DATA(className));
 		}
 	}
+	
 #endif /* JAVA_SPEC_VERSION >= 16 */
 
 restart:
diff --git a/runtime/vm/createramclass.cpp b/runtime/vm/createramclass.cpp
index 50950788c9..cc2a8ffac9 100644
--- a/runtime/vm/createramclass.cpp
+++ b/runtime/vm/createramclass.cpp
@@ -37,6 +37,8 @@
 #include "j9vmnls.h"
 #include "j2sever.h"
 #include "vm_internal.h"
+//inliningjclclasses : add include
+#include<iostream>
 
 #include "VMHelpers.hpp"
 
@@ -1982,15 +1984,56 @@ loadFlattenableFieldValueClasses(J9VMThread *currentThread, J9ClassLoader *class
 						* ImplicitCreation attribute. The attribute must have the ACC_DEFAULT flag set.
 						* Static fields will be checked during class preparation.
 						*/
+						if(!J9_IS_CLASSFILE_OR_ROMCLASS_VALUETYPE_VERSION(valueROMClass)){
+							//inliningjclclasses : getting and printing name of this romclass
+							J9UTF8 *classNameStructBH = J9ROMCLASS_CLASSNAME(valueROMClass);
+							std::cerr<<"THIS CLASS IS NULL RESTRICTED BUT NOT VALUETYPE_VERSION: ";
+							std::cerr.write(reinterpret_cast<const char*>(classNameStructBH->data), classNameStructBH->length);
+							std::cerr << std::endl;
+						}
+						if(!J9ROMCLASS_IS_VALUE(valueROMClass))
+						{
+							//inliningjclclasses : getting and printing name of this romclass
+							J9UTF8 *classNameStructBH = J9ROMCLASS_CLASSNAME(valueROMClass);
+							std::cerr<<"THIS CLASS IS NULL RESTRICTED, BUT NOT IS VALUE: ";
+							std::cerr.write(reinterpret_cast<const char*>(classNameStructBH->data), classNameStructBH->length);
+							std::cerr << std::endl;
+						}
+
+						if(J9_ARE_NO_BITS_SET(valueROMClass->optionalFlags, J9_ROMCLASS_OPTINFO_IMPLICITCREATION_ATTRIBUTE))
+						{
+							//inliningjclclasses : getting and printing name of this romclass
+							J9UTF8 *classNameStructBH = J9ROMCLASS_CLASSNAME(valueROMClass);
+							std::cerr<<"THIS CLASS IS NULL RESTRICTED, BUT NOT J9_ROMCLASS_OPTINFO_IMPLICITCREATION_ATTRIBUTE: ";
+							std::cerr.write(reinterpret_cast<const char*>(classNameStructBH->data), classNameStructBH->length);
+							std::cerr << std::endl;
+						}
+
+						if(J9_ARE_NO_BITS_SET(getImplicitCreationFlags(valueROMClass), J9AccImplicitCreateHasDefaultValue))
+						{
+							//inliningjclclasses : getting and printing name of this romclass
+							J9UTF8 *classNameStructBH = J9ROMCLASS_CLASSNAME(valueROMClass);
+							std::cerr<<"THIS CLASS IS NULL RESTRICTED, BUT NOT J9AccImplicitCreateHasDefaultValue: ";
+							std::cerr.write(reinterpret_cast<const char*>(classNameStructBH->data), classNameStructBH->length);
+							std::cerr << std::endl;
+						}
+
 						if (!J9ROMCLASS_IS_VALUE(valueROMClass)
 							|| J9_ARE_NO_BITS_SET(valueROMClass->optionalFlags, J9_ROMCLASS_OPTINFO_IMPLICITCREATION_ATTRIBUTE)
 							|| J9_ARE_NO_BITS_SET(getImplicitCreationFlags(valueROMClass), J9AccImplicitCreateHasDefaultValue)
 						) {
+
 							setCurrentExceptionForBadClass(currentThread, J9ROMCLASS_CLASSNAME(romClass), J9VMCONSTANTPOOL_JAVALANGINCOMPATIBLECLASSCHANGEERROR,
 								J9NLS_VM_NULLRESTRICTED_MUST_BE_IN_DEFAULT_IMPLICITCREATION_VALUE_CLASS);
 						}
 
 						if (!J9_IS_FIELD_FLATTENED(valueClass, field)) {
+							//inliningjclclasses : getting name of this romclass
+							J9UTF8 *classNameStructBH = J9ROMCLASS_CLASSNAME(valueROMClass);
+							std::cerr<<"THIS CLASS IS NULL RESTRICTED, BUT NOT J9_IS_FIELD_FLATTENED WITHIN THE CONTAINER CLASS: ";
+							std::cerr.write(reinterpret_cast<const char*>(classNameStructBH->data), classNameStructBH->length);
+							std::cerr << std::endl;
+
 							*valueTypeFlags |= (J9ClassContainsUnflattenedFlattenables | J9ClassHasReferences);
 							eligibleForFastSubstitutability = false;
 						} else if (J9_ARE_NO_BITS_SET(valueClass->classFlags, J9ClassCanSupportFastSubstitutability)) {
@@ -2325,6 +2368,13 @@ nativeOOM:
 #endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
 
 		if (J9ROMCLASS_IS_VALUE(romClass)) {
+			//inliningjclclasses : cerr print name of value class
+			
+			/*J9UTF8 *classNameStructBH = J9ROMCLASS_CLASSNAME(romClass);
+			std::cerr<<"INSIDE J9ROMCLASS_IS_VALUE() CONDITION WITHIN CREATERAMCLASS: ";
+			std::cerr.write(reinterpret_cast<const char*>(classNameStructBH->data), classNameStructBH->length);
+			std::cerr << std::endl;*/
+
 			classFlags |= J9ClassIsValueType;
 #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
 			/* superclass can be NULL if primitive classes are changed to value typess in the future. */
@@ -3461,9 +3511,11 @@ fail:
 					U_32 arrayFlags = J9ClassLargestAlignmentConstraintReference | J9ClassLargestAlignmentConstraintDouble;
 					if (J9_ARE_ALL_BITS_SET(options, J9_FINDCLASS_FLAG_CLASS_OPTION_NULL_RESTRICTED_ARRAY)) {
 						if (J9_ARE_ALL_BITS_SET(javaVM->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_ENABLE_VT_ARRAY_FLATTENING)) {
-							arrayFlags |= J9ClassIsFlattened;
+							//inliningjclclasses: commenting out this line
+							//arrayFlags |= J9ClassIsFlattened;
 						}
-						ramArrayClass->classFlags |= J9ClassArrayIsNullRestricted;
+						//inliningjclclasses : REMOVING THE SETTING OF NULLRESTRICTEDNESS 
+						//ramArrayClass->classFlags |= J9ClassArrayIsNullRestricted;
 					}
 					ramArrayClass->classFlags |= (elementClass->classFlags & arrayFlags);
 #endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
@@ -3493,7 +3545,8 @@ fail:
 					}
 				} else {
 					if (J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE(elementClass)) {
-						ramArrayClass->classFlags |= J9ClassContainsUnflattenedFlattenables;
+						//inliningjclclasses : commenting out this line
+						//ramArrayClass->classFlags |= J9ClassContainsUnflattenedFlattenables;
 					}
 					J9ARRAYCLASS_SET_STRIDE(ramClass, (((UDATA) 1) << (((J9ROMArrayClass*)romClass)->arrayShape & 0x0000FFFF)));
 				}
-- 
2.43.0

